File: simulator/backend/db/repositories/session-repository.ts
/**
 * Session Repository
 * Data access layer for session operations
 */

import { sql } from '../connection.ts';

export interface Session {
  id: string;
  user_id: string;
  token: string; // Changed from session_token
  expires_at: Date;
  created_at: Date;
}

export class SessionRepository {
  /**
   * Find session by token
   */
  async findByToken(token: string): Promise<Session | null> {
    try {
      const result = await sql`
        SELECT id, user_id, token, expires_at, created_at
        FROM sessions
        WHERE token = ${token} AND expires_at > NOW()
      `;
      return result[0] || null;
    } catch (error) {
      console.error('Error finding session by token:', error);
      throw error;
    }
  }

  /**
   * Create new session
   */
  async create(userId: string, token: string, expiresAt: Date, userEmail?: string): Promise<Session> {
    try {
      // Get user details if not provided
      let userName: string | undefined;
      let userRole: string | undefined;
      if (!userEmail) {
        const userResult = await sql`SELECT email, name, role FROM users WHERE id = ${userId}`;
        userEmail = userResult[0]?.email;
        userName = userResult[0]?.name;
        userRole = userResult[0]?.role;
      }

      const result = await sql`
        INSERT INTO sessions (user_id, user_email, user_name, user_role, token, expires_at)
        VALUES (${userId}, ${userEmail}, ${userName}, ${userRole}, ${token}, ${expiresAt})
        RETURNING id, user_id, token, expires_at, created_at
      `;
      return result[0];
    } catch (error) {
      console.error('Error creating session:', error);
      throw error;
    }
  }

  /**
   * Delete session by token
   */
  async deleteByToken(token: string): Promise<void> {
    try {
      await sql`
        DELETE FROM sessions
        WHERE session_token = ${token}
      `;
    } catch (error) {
      console.error('Error deleting session:', error);
      throw error;
    }
  }

  /**
   * Delete all sessions for user
   */
  async deleteAllForUser(userId: string): Promise<void> {
    try {
      await sql`
        DELETE FROM sessions
        WHERE user_id = ${userId}
      `;
    } catch (error) {
      console.error('Error deleting user sessions:', error);
      throw error;
    }
  }

  /**
   * Delete expired sessions
   */
  async deleteExpired(): Promise<void> {
    try {
      await sql`
        DELETE FROM sessions
        WHERE expires_at < NOW()
      `;
    } catch (error) {
      console.error('Error deleting expired sessions:', error);
      throw error;
    }
  }
}

// Export singleton instance
export const sessionRepository = new SessionRepository();

File: backend-python/app/repositories/session_repository.py
"""
Session Repository
Database operations for session entities
"""

import logging
from datetime import datetime
from uuid import uuid4

from app.database import fetch, fetchone, execute, record_to_dict, records_to_list
from app.models.session import SessionData

logger = logging.getLogger(__name__)


class SessionRepository:
    """Repository for session database operations."""

    async def find_by_token(self, token: str) -> SessionData | None:
        """Find session by token."""
        try:
            query = """
                SELECT id, user_id, token, expires_at, user_email, user_name, user_role, created_at
                FROM sessions WHERE token = $1 AND expires_at > NOW()
            """
            record = await fetchone(query, token)
            if not record:
                return None
            return SessionData(**record_to_dict(record))
        except Exception as e:
            logger.error(f"Error finding session by token: {e}")
            raise

    async def find_by_user_id(self, user_id: str) -> list[SessionData]:
        """Find all sessions for a user."""
        try:
            query = """
                SELECT id, user_id, token, expires_at, user_email, user_name, user_role, created_at
                FROM sessions WHERE user_id = $1 AND expires_at > NOW()
                ORDER BY created_at DESC
            """
            records = await fetch(query, user_id)
            return [SessionData(**r) for r in records_to_list(records)]
        except Exception as e:
            logger.error(f"Error finding sessions by user ID: {e}")
            raise

    async def create(
        self,
        user_id: str,
        token: str,
        expires_at: datetime,
        user_email: str | None = None,
        user_name: str | None = None,
        user_role: str | None = None,
    ) -> SessionData:
        """Create a new session."""
        try:
            session_id = str(uuid4())

            # If user info not provided, fetch from users table
            if not user_email:
                user_query = "SELECT email, name, role FROM users WHERE id = $1"
                user_record = await fetchone(user_query, user_id)
                if user_record:
                    user_dict = record_to_dict(user_record)
                    user_email = user_dict.get('email')
                    user_name = user_dict.get('name')
                    user_role = user_dict.get('role')

            query = """
                INSERT INTO sessions (user_id, user_email, user_name, user_role, token, expires_at)
                VALUES ($1, $2, $3, $4, $5, $6)
                RETURNING id, user_id, token, expires_at, user_email, user_name, user_role, created_at
            """
            record = await fetchone(
                query,
                user_id,
                user_email,
                user_name,
                user_role,
                token,
                expires_at,
            )
            return SessionData(**record_to_dict(record))
        except Exception as e:
            logger.error(f"Error creating session: {e}")
            raise

    async def delete_by_token(self, token: str) -> None:
        """Delete session by token."""
        try:
            query = "DELETE FROM sessions WHERE token = $1"
            await execute(query, token)
        except Exception as e:
            logger.error(f"Error deleting session by token: {e}")
            raise

    async def delete_all_for_user(self, user_id: str) -> None:
        """Delete all sessions for a user."""
        try:
            query = "DELETE FROM sessions WHERE user_id = $1"
            await execute(query, user_id)
        except Exception as e:
            logger.error(f"Error deleting sessions for user: {e}")
            raise

    async def delete_expired(self) -> int:
        """Delete all expired sessions."""
        try:
            query = "DELETE FROM sessions WHERE expires_at <= NOW()"
            result = await execute(query)
            # Extract count from result string like "DELETE 5"
            if result and "DELETE" in result:
                parts = result.split()
                if len(parts) >= 2:
                    return int(parts[1])
            return 0
        except Exception as e:
            logger.error(f"Error deleting expired sessions: {e}")
            raise

    async def update_expiry(self, token: str, expires_at: datetime) -> None:
        """Update session expiry."""
        try:
            query = "UPDATE sessions SET expires_at = $2 WHERE token = $1"
            await execute(query, token, expires_at)
        except Exception as e:
            logger.error(f"Error updating session expiry: {e}")
            raise

    async def count_active_sessions(self, user_id: str) -> int:
        """Count active sessions for a user."""
        try:
            query = "SELECT COUNT(*) FROM sessions WHERE user_id = $1 AND expires_at > NOW()"
            result = await fetchone(query, user_id)
            return result[0] if result else 0
        except Exception as e:
            logger.error(f"Error counting active sessions: {e}")
            raise


# Singleton instance
session_repository = SessionRepository()
